### ISSUE TITLE
TorchScript gelu signature difference

### ISSUE THREAD

**Issue by leeho** [triton-inference-server/server#3385]
I am trying to run the SentenceTransformer model from sentence-transformers==0.2.6.2 on Triton. As the model itself contains quite some logic, I exported a part of it - BertEncoder ( https://github.com/huggingface/transformers/blob/v2.11.0/src/transformers/modeling_bert.py#L386 ) - as a TorchScript, as described here: https://medium.com/nvidia-ai/how-to-deploy-almost-any-hugging-face-model-on-nvidia-triton-inference-server-with-an-8ee7ec0e6fc4

I can run the exported TorchScript with torch just fine, but when I deploy this model on triton, I get the following error on model load:

```
E0921 17:03:26.391509 1 model_repository_manager.cc:1215] failed to load 'transformer' version 1: Internal: failed to load model 'transformer': 
Arguments for call are not valid.
The following variants are available:
  
  aten::gelu(Tensor self, bool approximate) -> (Tensor):
  Argument approximate not provided.
  
  aten::gelu.out(Tensor self, bool approximate, *, Tensor(a!) out) -> (Tensor(a!)):
  Argument approximate not provided.

The original call is:
/home/oleyba/miniconda3/envs/sentence_transfomer_serving/lib/python3.7/site-packages/torch/nn/functional.py(1555): gelu
/home/oleyba/code/misc/sent_transformer/test_with_tracing_encoder.py(44): forward
/home/oleyba/miniconda3/envs/sentence_transfomer_serving/lib/python3.7/site-packages/torch/nn/modules/module.py(1039): _slow_forward
/home/oleyba/miniconda3/envs/sentence_transfomer_serving/lib/python3.7/site-packages/torch/nn/modules/module.py(1051): _call_impl
/home/oleyba/miniconda3/envs/sentence_transfomer_serving/lib/python3.7/site-packages/transformers/modeling_bert.py(380): forward
/home/oleyba/miniconda3/envs/sentence_transfomer_serving/lib/python3.7/site-packages/torch/nn/modules/module.py(1039): _slow_forward
/home/oleyba/miniconda3/envs/sentence_transfomer_serving/lib/python3.7/site-packages/torch/nn/modules/module.py(1051): _call_impl
/home/oleyba/code/misc/sent_transformer/test_with_tracing_encoder.py(85): forward
/home/oleyba/miniconda3/envs/sentence_transfomer_serving/lib/python3.7/site-packages/torch/nn/modules/module.py(1039): _slow_forward
/home/oleyba/miniconda3/envs/sentence_transfomer_serving/lib/python3.7/site-packages/torch/nn/modules/module.py(1051): _call_impl
/home/oleyba/miniconda3/envs/sentence_transfomer_serving/lib/python3.7/site-packages/torch/jit/_trace.py(959): trace_module
/home/oleyba/miniconda3/envs/sentence_transfomer_serving/lib/python3.7/site-packages/torch/jit/_trace.py(744): trace
/home/oleyba/code/misc/sent_transformer/test_with_tracing_encoder.py(178): torch_trace
/home/oleyba/code/misc/sent_transformer/test_with_tracing_encoder.py(234): forward
/home/oleyba/miniconda3/envs/sentence_transfomer_serving/lib/python3.7/site-packages/torch/nn/modules/module.py(1051): _call_impl
/home/oleyba/miniconda3/envs/sentence_transfomer_serving/lib/python3.7/site-packages/sentence_transformers/models/Transformer.py(25): forward
/home/oleyba/miniconda3/envs/sentence_transfomer_serving/lib/python3.7/site-packages/torch/nn/modules/module.py(1051): _call_impl
/home/oleyba/miniconda3/envs/sentence_transfomer_serving/lib/python3.7/site-packages/torch/nn/modules/container.py(139): forward
/home/oleyba/miniconda3/envs/sentence_transfomer_serving/lib/python3.7/site-packages/sentence_transformers/SentenceTransformer.py(150): encode
/home/oleyba/code/misc/sent_transformer/test_with_tracing_encoder.py(244): <module>
Serialized   File "code/__torch__.py", line 37
  def forward(self: __torch__.MyBertIntermediate,
    argument_1: Tensor) -> Tensor:
    input = torch.gelu((self.dense).forward(argument_1, ))
            ~~~~~~~~~~ <--- HERE
    return input
```
Indeed, when I check gelu signatures, they do not match, this is the original gelu implementation from pytorch:
( https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L1544 )

```
def gelu(input):
    r"""gelu(input) -> Tensor

    Applies element-wise the function
    :math:`\text{GELU}(x) = x * \Phi(x)`

    where :math:`\Phi(x)` is the Cumulative Distribution Function for Gaussian Distribution.

    See `Gaussian Error Linear Units (GELUs) <https://arxiv.org/abs/1606.08415>`_.
    """
    if has_torch_function_unary(input):
        return handle_torch_function(gelu, (input,), input)
    return torch._C._nn.gelu(input)
```

and this is an implementation I copied from the PyTorch NGC container nvcr.io/nvidia/pytorch:21.08-py3 as suggested here (https://github.com/triton-inference-server/pytorch_backend#build-the-pytorch-backend-with-custom-pytorch)

```
def gelu(input: Tensor, approximate: bool = False) -> Tensor:
    r"""gelu(input, approximate) -> Tensor

    Applies element-wise the function
    :math:`\text{GELU}(x) = x * \Phi(x)`

    where :math:`\Phi(x)` is the Cumulative Distribution Function for Gaussian Distribution.

    When the approximate flag is enabled, Gelu is estimated with:
        :math::  \text{GELU}(x) = 0.5 * x * (1 + \text{Tanh}(\sqrt(2 / \pi) * (x + 0.044715 * x^3)))

    See `Gaussian Error Linear Units (GELUs) <https://arxiv.org/abs/1606.08415>`_.
    """
    if has_torch_function_unary(input):
        return handle_torch_function(gelu, (input,), input, approximate=approximate)
    return torch._C._nn.gelu(input, approximate)
```

I see that the new parameter has a default value, but the model still does not load. Could you please advise how this model can be exported, or is this behavior a bug?

**Comment by CoderHam** [comment]
> I can run the exported TorchScript with torch
Did you use the version of PyTorch from the 21.08 NGC container? Did you use the C or the Python API?

Which pytorch version / PyTorch NGC container version was used for producing the model?

**Comment by leeho** [comment]
Thanks a lot for your answer, @CoderHam , indeed my mistake was that I traced the model on my local environment. I have now traced it again in the NGC container nvcr.io/nvidia/pytorch:21.08-py3 and this traced model now loads correctly on triton.

I see that the output of the model after tracing is now a little different compared to running the original one with pytorch, even from within the NGC container - i.e. if I run prediction with SentenceTransformer from the container, save results and then run again with the BertEncoder loaded from TorchScript within the same container (without triton). For most cases, the difference is minor. This was not the case when I traced the model locally and tested with loading this model as TorchScript locally. This should not be the issue for our case, but I will leave this here as an additional information.
